### ISSUE TITLE
Strategy for dealing with `as u64` casts

### ISSUE THREAD

**Issue by apoelstra** [rust-bitcoin/rust-bitcoin#2941]
>So what? If it defines a trait that should be potentially public and global for the ecosystem

Maybe. It looks like the usage in #2929 comes from two places: one is 

> I only partially agree here. It's supposed to be used in `WriteExt` method signature which makes it part of public API.

For this purpose, we could define it alongside `WriteExt` and call it `VarintEncodable` or something.

Looking at #2929, it seems we actually need `ToU64` in a diverse set of places *but* I think that almost all of them can be eliminated with other strategies:

* In BIP158, where we need it because we're storing indices in a `Vec<u64>`; we should just change that to a `Vec<usize>`. Presumably the original code was done this way to make interop with `VarInt` easier and more obviously correct.
* To call `Weight::from_non_witness_data_size`, which again is using a u64 as a size, and should probably just take `usize` (same story with `Weight::from_vb` ad `FeeRate::from_vb`)
* In `minimal_non_dust_inner` which is an internal function legitimately converting a usize transaction size to a u64 fee amount, where maybe we can live with an `expect`.
* In `push_slice_no_opt` where ironically, we explicitly panic on values that aren't convertible to u32s, so we should just be using `u32::try_from`.
* In the encoding module, where much of the time `VarintEncodable` will do what we want. We are also converting `MAX_MSG_SIZE` and `MAX_VEC_SIZE` to u64s to pass into `io::Read::Take` but here we're doing range-checks anyway so we could just use `try_from` and propagate the error since any failed conversion to u64 or even u32 means a size is out of bounds. Maybe we want an internal `ReaderExt` trait or something that would make this more convenient for us. Conversely `Cursor::position` is a u64 but we want to interpret it as a usize. Again maybe `expect` would be fine here, because it only occurs a couple times.

_Originally posted by @apoelstra in https://github.com/rust-bitcoin/rust-bitcoin/discussions/2938#discussioncomment-9916823_

**Comment by Kixunil** [comment]
> and call it `VarintEncodable` or something.

Yeah, this is almost `Encode` from push_decode PR but it needs to explicitly say varint (compact size).

I could kind of understand `Weight` (sadly it's still not overflow-free because of segwit discount) but there are issues with `FeeRate` since it's computed as `Amount / Weight` and `Amount` definitely has to be `u64`. So what should `FeeRate` be?

> * In `push_slice_no_opt` where ironically, we explicitly panic on values that aren't convertible to u32s, so we should just be using `u32::try_from`

We can also just operate on `usize` to maintain the nice `match` structure.

> * `MAX_MSG_SIZE` and `MAX_VEC_SIZE`

Those are guaranteed to fit `u32` which is convertible to `u64` and `usize` because we don't support < 32-bit architectures. I want a `ToUsize` trait in `internals` that is analogous to `ToU64`.

**Comment by tcharding** [comment]
I've added PR links to the bullet points that I addressed. The rest need addressing or a counterargument before this issue can be closed.

**Comment by Kixunil** [comment]
* For `Weight` and `FeeRate` we still need conversion to happen *somewhere* because `Amount` is `u64` by consensus. So I'd pick `u64` for the inner types and do the conversion inside `Weight::from_non_witness_data_size` which indeed should use `usize`.
* `minimal_non_dust_inner` was renamed to `minimal_non_dust_internal` and we've changed it to use `to_u64`
* When converting `MAX_MSG_SIZE` and `MAX_VEC_SIZE` to u64s to pass into `io::Read::Take` I think `to_u64` is appropriate
* `Cursor::position` is in `deserialize_partial` function which should've just taken `&mut &[u8]` anyway but `Cursor` isn't needed either. Just remember the original length of the slice, pass the slice to the decoding function and then subtract the resulting slice from the original one.
